src/pinns_pde_solve.jl

```diff
@@ -282,80 +293,49 @@ function generate_training_sets(domains,dx,bcs,indvars,depvars,dict_indvars,dict
-        push!(train_set, [points...])
-    end
+    train_set = map(points -> [points...], Iterators.product(spans...))
 
-    train_bound_set = []
-    for bt in bound_args
-        _set = []
-        span = [get(dict_var_span, b, b) for b in bt]
-        for points in Iterators.product(span...)
-            push!(_set, [points...])
-        end
-        push!(train_bound_set, _set)
+    train_bound_set = map(bound_args) do bt
+        span = map(b -> get(dict_var_span, b, [b]), bt)
+        _set = map(points -> [points...], Iterators.product(span...))
     end
 
-    flat_train_bound_set = collect(Iterators.flatten(train_bound_set))
+    flat_train_bound_set = map(x -> x, Iterators.flatten(train_bound_set))
     train_domain_set =  setdiff(train_set, flat_train_bound_set)
 
     [train_domain_set,train_bound_set,train_set]
 end
 
-function get_phi(chain,isuinplace)
+function get_phi(chain)
+    chain isa Chain && "Chain deprecated, use FastChain"
+    isuinplace = chain.layers[end].out == 1
     # The phi trial solution
-    if chain isa FastChain
-        if isuinplace
-            phi = (x,θ) -> first(chain(adapt(typeof(θ),x),θ))
-        else
-            phi = (x,θ) -> chain(adapt(typeof(θ),x),θ)
-        end
+    if isuinplace
+        phi = (x,θ) -> first(chain(adapt(typeof(θ),x),θ))
     else
-        _,re  = Flux.destructure(chain)
-        if isuinplace
-            phi = (x,θ) -> first(re(θ)(adapt(typeof(θ),x)))
-        else
-            phi = (x,θ) -> re(θ)(adapt(typeof(θ),x))
-        end
+        phi = (x,θ) -> chain(adapt(typeof(θ),x),θ)
     end
     phi
 end
 
-function get_derivative(dim,phi,autodiff,isuinplace)
+# the method  calculate derivative
+function get_derivative(autodiff, isuinplace)
     epsilon = cbrt(eps(Float32))
-     function get_ε(dim, der_num)
-         ε = zeros(dim)
-         ε[der_num] = epsilon
-         ε
-     end
-     εs = [get_ε(dim,d) for d in 1:dim]
-
     if autodiff # automatic differentiation (not implemented yet)
         error("automatic differentiation is not implemented yet)")
     else # numerical differentiation
-        derivative = (_x...) ->
+        derivative = (u,var_num,x,εs,order,θ) ->
         begin
-            var_num = _x[1]
-            x = collect(_x[2:end-3])
-            dnv =_x[end-2]
-            order = _x[end-1]
-            θ = _x[end]
-            εs_dnv = [εs[d] for d in dnv]
-            return _derivative(x,θ,order,εs_dnv,var_num)
-        end
-        _derivative = (x,θ,order,εs_dnv,var_num) ->
-        begin
-            ε = εs_dnv[order]
+            ε = εs[order]
             if order == 1
                 if isuinplace
-                    return (phi(x+ε,θ) - phi(x-ε,θ))/(2*epsilon)
+                    return (u(x+ε,θ) - u(x-ε,θ))/(2*epsilon)
                 else
-                    return (phi(x+ε,θ)[var_num] - phi(x-ε,θ)[var_num])/(2*epsilon)
+                    return (u(x+ε,θ)[var_num] - u(x-ε,θ)[var_num])/(2*epsilon)
```

Comment:
you're going to repeat derivative calls. The computation should be re-ordered next PR.